contracts/Pool.sol

```diff
@@ -28,3 +28,84 @@ contract Pool {
         receiver.transfer(value);
     }
 }
+
+struct ReceiverWeight {
+    address next;
+    uint32 weight;
+}
+
+/// @notice Helper methods for receiver weights list.
+/// The list works optimally if after applying a series of changes it's iterated over.
+/// The list uses 1 word of storage per receiver with a non-zero weight.
+library ReceiverWeightsImpl {
+    address internal constant ADDR_ROOT = address(0);
+    address internal constant ADDR_UNINITIALIZED = address(0);
+    address internal constant ADDR_END = address(1);
+
+    function init(mapping(address => ReceiverWeight) storage self) internal {
+        self[ADDR_ROOT].next = ADDR_END;
+    }
+
+    /// @notice Return the next non-zero receiver weight and its address.
+    /// Removes all the zeroed items found between the current and the next receivers.
+    /// Iterating over the whole list removes all the zeroed items.
+    /// @param current The previously returned receiver address or zero to start iterating
+    /// @return next The next receiver address
+    /// @return weight The next receiver weight, zero if the end of the list was reached
+    function nextWeight(
+        mapping(address => ReceiverWeight) storage self,
+        address current
+    ) internal returns (address next, uint32 weight) {
+        ReceiverWeight storage currentReceiverWeight = self[current];
```

Comment:
I couldn't find much information about the optimizer.

Would it cost anything if I replaced every usage of `currentReceiverWeight` with `self[current]`? Key hashing cost is real, but does the optimizer prevent its recalculation?

Reply:
It should be easy to run the tests and see how much gas is used in either case.

Reply:
I was looking in ethers.io docs for functionality like this, but I couldn't find it, do you have a tip up your sleeve? If not, I'll look harder.

Reply:
It's optimized

Reply:
`TransactionReceipt::gasUsed()` will give you the amount. We can use that to optimize later.

Reply:
Thanks :)